boxsdk/auth/oauth2.py

```diff
@@ -320,3 +342,72 @@ def revoke(self):
             if not network_response.ok:
                 raise BoxOAuthException(network_response.status_code, network_response.content, url, 'POST')
             self._store_tokens(None, None)
+
+    def close(self, revoke=True):
+        """Close the auth object.
+
+        After this action is performed, the auth object can no longer request
+        new tokens.
+
+        This method may be called even if the auth object is already closed.
+
+        :param revoke:
+            (optional) Whether the current tokens should be revoked, via `revoke()`.
+            Defaults to `True` as a security precaution, so that the tokens aren't usable
+            by any adversaries after you are done with them.
+            Note that the revoke isn't guaranteed to succeed (the network connection might
+            fail, or the API call might respond with a non-200 HTTP response), so this
+            isn't a fool-proof security mechanism.
+            If the revoke fails, an exception is raised.
+            The auth object is still considered to be closed, even if the revoke fails.
+        :type revoke:   `bool`
+        """
+        self._closed = True
+        if revoke:
+            self.revoke()
+
+    @contextmanager
+    def closing(self, **close_kwargs):
+        """Context manager to close the auth object on exit.
+
+        The behavior is somewhat similar to `contextlib.closing(self)`, but has
+        some differences.
+
+        The context manager cannot be entered if the auth object is closed.
+
+        If a non-`Exception` (e.g. `KeyboardInterrupt`) is caught from the
+        block, this context manager prioritizes re-raising the exception as
+        fast as possible, without blocking. Thus, in this case, the tokens will
+        not be revoked, even if `revoke=True` was passed to this method.
+
+        If exceptions are raised both from the block and from `close()`, the
+        exception from the block will be reraised, and the exception from
+        `close()` will be swallowed. The assumption is that the exception from
+        the block is more relevant to the client, especially since the revoke
+        can fail if the network is unavailable.
+
+        :param **close_kwargs:  Keyword arguments to pass to `close()`.
+        """
+        self._check_closed()
+        exc_infos = []
+
+        # pylint:disable=broad-except
+        try:
+            yield self
+        except Exception:
+            exc_infos.append(sys.exc_info())
+        except BaseException:
+            exc_infos.append(sys.exc_info())
+            close_kwargs['revoke'] = False
+
+        try:
+            self.close(**close_kwargs)
+        except Exception:
+            exc_infos.append(sys.exc_info())
+
+        if exc_infos:
+            six.reraise(*exc_infos[0])
+
+    def _check_closed(self):
+        if self.closed:
+            raise ValueError("operation on a closed auth object")
```

Comment:
I'm not sure `ValueError` is correct here. I'd raise `RuntimeError` or a custom `Exception`